build/emoji/emoji_gen.mjs

```diff
@@ -0,0 +1,273 @@
+
+import jsonData from 'emoji-datasource/emoji.json';
+import jsonCategories from 'emoji-datasource/categories.json';
+
+const EMOJI_SIZE = 64;
+const endResults = [];
+
+// copy image files
+const source = `node_modules/emoji-datasource-apple/img/apple/${EMOJI_SIZE}/`;
+const readDirPromise = Fs.readdir(source);
+endResults.push(readDirPromise);
+readDirPromise.then((images) => {
+    console.log(`I will now copy ${images.length} emoji images, this might take a while`);
+    for (const imageFile of images) {
+        endResults.push(Fs.copyFile(
+            path.join(source, imageFile),
+            path.join('images/emoji', imageFile),
+        ).catch((err) => console.log(`[ERROR] Failed to copy ${imageFile}: ${err}`)));
+    }
+});
+
+// copy sheet image
+const sheetSource = `node_modules/emoji-datasource-apple/img/apple/sheets/${EMOJI_SIZE}.png`;
+const sheetFile = 'images/emoji-sheets/apple-sheet.png';
+console.log('Copy sprite sheet');
+Fs.copyFile(sheetSource, sheetFile).catch((err) => console.log(`[ERROR] Failed to copy sheet file: ${err}`));
+
+// we'll load it as a two dimensional array so we can generate a Map out of it
+const emojiIndicesByAlias = [];
+const emojiIndicesByUnicode = [];
+const emojiIndicesByCategory = new Map();
+const categoryNamesSet = new Set();
+const categoryDefaultTranslation = new Map();
+const emojiImagesByAlias = [];
+const emojiFilePositions = new Map();
+const skinCodes = {
+    '1F3FB': 'light_skin_tone',
+    '1F3FC': 'medium_light_skin_tone',
+    '1F3FD': 'medium_skin_tone',
+    '1F3FE': 'medium_dark_skin_tone',
+    '1F3FF': 'dark_skin_tone',
+};
+
+const skinNames = {
+    '1F3FB': 'LIGHT SKIN TONE',
+    '1F3FC': 'MEDIUM LIGHT SKIN TONE',
+    '1F3FD': 'MEDIUM SKIN TONE',
+    '1F3FE': 'MEDIUM DARK SKIN TONE',
+    '1F3FF': 'DARK SKIN TONE',
+};
+const control = new AbortController();
+const writeOptions = {
+    encoding: 'utf8',
+    signal: control.signal,
+};
+
+function filename(emoji) {
+    return emoji.image.split('.')[0];
+}
+
+function writeFile(fileName, filePath, data) {
+    const promise = Fs.writeFile(filePath, data, writeOptions);
+
+    promise.then(() => {
+        console.log(`${fileName} generated successfuly.`);
+    });
+    return promise;
+}
+
+function convertCategory(category) {
+    return category.toLowerCase().replace(' & ', '-');
+}
+
+function genSkinVariations(emoji, index, nextOrder) {
+    if (!emoji.skin_variations) {
+        return [];
+    }
+    let order = nextOrder;
+    return Object.keys(emoji.skin_variations).map((skinCode) => {
+        // if skin codes ever change this will produce a null_light_skin_tone
+        const skins = skinCode.split('-');
+        const skinShortName = skins.map((code) => skinCodes[code]).join('_');
+        const skinName = skins.map((code) => skinNames[code]).join(', ');
+        const variation = {...emoji.skin_variations[skinCode]};
+        variation.short_name = `${emoji.short_name}_${skinShortName}`;
```

Comment:
What are the changes to the format of the emoji.json file? I think Matterhorn uses a copy of it to do their emoji replacement, so I want to be able to let them know what changes there'll be before we actually ship them

Reply:
`name` is not the usual `my_emoji`, but an angry representation of the same one `MY EMOJI`, for that you should use `short_name`
there is no `alias` but `short_names`, which includes the canonical `short_name` in its first position
instead of filename there is `image` which also contains the file extension

if there are skin variations, it has a field `skin_variations` which contains the needed info for that variation, but in our script we convert those to full emojis and add them (in case they are using `utils/emoji.json`, if they use the one provided by the library you can disregard this), I also aded a skins property which is an array of skin codes, I use the first one (if there are more than one) as the one to filter by

there are other fields included that we are not using, but it might be useful to someone so I left them, you can find more info in the [lib's readme](https://github.com/iamcal/emoji-data#using-the-data)

Reply:
Cool, I'll let them know then. It seems like the two big changes that would probably affect them are renaming aliases to short_names and possibly the change to filename, but I can't imagine either of those are that big